[Lens] Display heatmap on suggestions

stratoula opened the pull request:
## Summary

As heatmap is out of experimental we can now allow it to be displayed on the suggestions panel

![image](https://user-images.githubusercontent.com/17003240/152373304-0e71adf8-7946-4b61-b78a-9761faa24183.png)

### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

stratoula commented:
@elasticmachine merge upstream

mbondyra commented:
Hmmm, I think we should hide heatmap suggestion when there's a date histogram in the visualization. What do you think?
<img width="975" alt="Screenshot 2022-02-07 at 10 55 37" src="https://user-images.githubusercontent.com/4283304/152765588-0edb5928-c223-45a7-8b8e-39970ecd36c2.png">

mbondyra reviewed (commented):
I think we should hide heatmap suggestion when there's a date histogram in the visualization. 
I think the only valuable suggestions would be generated when having two non-date-histogram bucket groups 🤔 

The two below feel not very good charts:
<img width="500" alt="Screenshot 2022-02-07 at 10 55 37" src="https://user-images.githubusercontent.com/4283304/152765588-0edb5928-c223-45a7-8b8e-39970ecd36c2.png">
<img width="500" alt="Screenshot 2022-02-07 at 11 00 11" src="https://user-images.githubusercontent.com/4283304/152766299-cd51da75-22db-40a6-bc74-0fdeef00fa51.png">

Essentially I'd replace the `hide` value to:
```
/**
   * Hide for:
   * - reduced and reorder tables
   * - tables without 2 non-date-histogram buckets
   */
  const isTwoNonDateHistogramBuckets = groups.length === 2 && groups.every(group => group.operation.dataType !== 'date') && metrics.length == 1;

  const hide =
    table.changeType === 'reduced' || table.changeType === 'reorder' || !isTwoNonDateHistogramBuckets;
```

What do you think?

flash1293 commented:
I agree @mbondyra , date histogram + metric doesn't make much sense as heatmap. However I don't think we need two non-date histogram bucket dimensions, one seems enough:
<img width="679" alt="Screenshot 2022-02-07 at 12 05 43" src="https://user-images.githubusercontent.com/1508364/152776672-e35e5b16-2bb4-4405-8991-913cd9fbda00.png">

@stratoula I know it's not the focus of this PR, but I noticed the heatmap suggestion is somehow rendered a bit off:
<img width="501" alt="Screenshot 2022-02-07 at 12 02 45" src="https://user-images.githubusercontent.com/1508364/152776371-edb9c8fd-fdaa-484f-971e-46fb5470b0de.png">

It's just taking half of the available space. Is this an elastic-charts bug we should report?

stratoula commented:
@flash1293 @mbondyra great feedback thanx. I have updated the PR and I am hiding the heatmap suggestions if date histogram is applied.

@flash1293 u r right. This happens only when I have a date histogram applied, in all other cases it renders well. I have pinged @markov00 offline to discuss it.

As it happens only for date-histogram and we are hiding it now, I think is it fine.

flash1293 commented:
@stratoula as I noted I think we should show date histogram based suggestions as long as there is a second bucket dimension.

stratoula commented:
Sorry I got confused with the comments, I hope I got it correctly now.

mbondyra commented on x-pack/plugins/lens/public/heatmap_visualization/suggestions.ts:
```diff
@@ -58,9 +58,17 @@ export const getSuggestions: Visualization<HeatmapVisualizationState>['getSugges
    * - tables with just a single bucket dimension
+   * - tables with a date histogram
    */
+  const hasOnlyOneDatehistogramBucket =
+    groups.some((group) => group.operation.dataType === 'date') &&
+    metrics.length === 1 &&
+    groups.length === 1;
```
let's discuss this condition during sync, but if we decide to stick to this solution, I think it'll be better to change the order of execution and check if all groups are date_histogram:

```suggestion
  const hasOnlyDatehistogramBuckets =
    metrics.length === 1 &&
    groups.length > 0 &&
    groups.every((group) => group.operation.dataType === 'date')
```

stratoula commented:
@flash1293 about the heatmap suggestions bug
1. we opened an issue here https://github.com/elastic/elastic-charts/issues/1581
2. I fixed it with a bit hacky way :D 

![image](https://user-images.githubusercontent.com/17003240/152812258-3e580cdc-d6c2-466d-9997-280e098bada1.png)

It renders fine now. I have also changed the score for the heatmaps with date histogram, so it appears on the right now. 

Let me know if you are fine with it :)

flash1293 reviewed (approved):
Tested and works great now, thanks!

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.1|**Backport failed because of merge conflicts**|

### How to fix
Re-run the backport manually:
```
node scripts/backport --pr 124542
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)
